File: step_definitions/Real-Time Shipment Status Updates.java
package com.cucumber.steps;

import com.page_objects.ShipmentTrackingPage;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

public class ShipmentTrackingSteps extends ShipmentTrackingPage {

    @Given("User is logged into the system")
    public void userIsLoggedIntoTheSystem() {
        loginToSystem();
    }

    @When("User navigates to the shipment tracking page")
    public void userNavigatesToTheShipmentTrackingPage() {
        navigateToShipmentTrackingPage();
    }

    @Then("Shipment tracking page is displayed")
    public void shipmentTrackingPageIsDisplayed() {
        verifyShipmentTrackingPageDisplayed();
    }

    @Then("Shipment data is available in the system")
    public void shipmentDataIsAvailableInTheSystem() {
        verifyShipmentDataAvailable();
    }

    @Given("Shipment ID is \"12345\"")
    public void shipmentIDIs() {
        setShipmentID("12345");
    }

    @When("User enters the Shipment ID \"12345\" in the tracking field")
    public void userEntersTheShipmentIDInTheTrackingField() {
        enterShipmentID("12345");
    }

    @Then("Shipment details for ID \"12345\" are displayed")
    public void shipmentDetailsForIDAreDisplayed() {
        verifyShipmentDetailsDisplayed("12345");
    }

    @Given("User ID is \"67890\"")
    public void userIDIs() {
        setUserID("67890");
    }

    @When("User checks the current status of the shipment")
    public void userChecksTheCurrentStatusOfTheShipment() {
        checkCurrentShipmentStatus();
    }

    @Then("Current status is displayed as \"In Transit\"")
    public void currentStatusIsDisplayedAs() {
        verifyCurrentStatus("In Transit");
    }

    @When("User simulates a status update to \"Out for Delivery\"")
    public void userSimulatesAStatusUpdateToOutForDelivery() {
        simulateStatusUpdate("Out for Delivery");
    }

    @Then("Status updates to \"Out for Delivery\" in real-time")
    public void statusUpdatesToOutForDeliveryInRealTime() {
        verifyStatusUpdate("Out for Delivery");
    }

    @When("User simulates a status update to \"Delivered\"")
    public void userSimulatesAStatusUpdateToDelivered() {
        simulateStatusUpdate("Delivered");
    }

    @Then("Status updates to \"Delivered\" in real-time")
    public void statusUpdatesToDeliveredInRealTime() {
        verifyStatusUpdate("Delivered");
    }

    @When("User verifies the timestamp of the latest status update")
    public void userVerifiesTheTimestampOfTheLatestStatusUpdate() {
        verifyTimestampOfLatestStatusUpdate();
    }

    @Then("Timestamp matches the current time")
    public void timestampMatchesTheCurrentTime() {
        verifyTimestampMatchesCurrentTime();
    }

    @When("User refreshes the page")
    public void userRefreshesThePage() {
        refreshPage();
    }

    @Then("Status remains \"Delivered\" with the same timestamp")
    public void statusRemainsDeliveredWithTheSameTimestamp() {
        verifyStatusAndTimestamp("Delivered");
    }

    @When("User logs out and logs back in")
    public void userLogsOutAndLogsBackIn() {
        logoutAndLogin();
    }

    @Then("Shipment status remains \"Delivered\"")
    public void shipmentStatusRemainsDelivered() {
        verifyShipmentStatus("Delivered");
    }

    @When("User checks the notification settings for shipment updates")
    public void userChecksTheNotificationSettingsForShipmentUpdates() {
        checkNotificationSettings();
    }

    @Then("Notifications are enabled for real-time updates")
    public void notificationsAreEnabledForRealTimeUpdates() {
        verifyNotificationsEnabled();
    }

    @When("User simulates a network issue and attempts to update status")
    public void userSimulatesANetworkIssueAndAttemptsToUpdateStatus() {
        simulateNetworkIssueAndUpdateStatus();
    }

    @Then("System handles the issue gracefully and retries the update")
    public void systemHandlesTheIssueGracefullyAndRetriesTheUpdate() {
        verifySystemHandlesNetworkIssue();
    }

    @When("User verifies the shipment history log")
    public void userVerifiesTheShipmentHistoryLog() {
        verifyShipmentHistoryLog();
    }

    @Then("Log contains all status updates with timestamps")
    public void logContainsAllStatusUpdatesWithTimestamps() {
        verifyLogContainsStatusUpdates();
    }

    @When("User checks for any error messages during status updates")
    public void userChecksForAnyErrorMessagesDuringStatusUpdates() {
        checkForErrorMessages();
    }

    @Then("No error messages are displayed")
    public void noErrorMessagesAreDisplayed() {
        verifyNoErrorMessages();
    }

    @When("User attempts to update status from a different device")
    public void userAttemptsToUpdateStatusFromADifferentDevice() {
        attemptStatusUpdateFromDifferentDevice();
    }

    @Then("Status updates are synchronized across devices")
    public void statusUpdatesAreSynchronizedAcrossDevices() {
        verifyStatusSynchronizationAcrossDevices();
    }

    @When("User verifies the shipment status on a mobile device")
    public void userVerifiesTheShipmentStatusOnAMobileDevice() {
        verifyShipmentStatusOnMobileDevice();
    }

    @Then("Status is consistent with the desktop view")
    public void statusIsConsistentWithTheDesktopView() {
        verifyStatusConsistencyWithDesktopView();
    }

    @When("User checks the shipment status after a system reboot")
    public void userChecksTheShipmentStatusAfterASystemReboot() {
        checkShipmentStatusAfterSystemReboot();
    }

    @Then("Status remains \"Delivered\"")
    public void statusRemainsDelivered() {
        verifyStatusAfterReboot("Delivered");
    }
}

File: step_definitions/Real-Time Shipment Status Updates.js
import { Given, When, Then } from "cypress-cucumber-preprocessor/steps";
import ShipmentTrackingPage from "../pages/shipmentTrackingPage";

Given("the user is logged into the system", () => {
 cy.login();
});

When("the user navigates to the shipment tracking page", () => {
 ShipmentTrackingPage.visit();
});

Then("the shipment tracking page is displayed", () => {
 ShipmentTrackingPage.verifyPageDisplayed();
});

Given("the shipment tracking page is displayed", () => {
 ShipmentTrackingPage.verifyPageDisplayed();
});

When("the user enters the Shipment ID {string} in the tracking field", (shipmentId) => {
 ShipmentTrackingPage.enterShipmentID(shipmentId);
});

Then("shipment details for ID {string} are displayed", (shipmentId) => {
 ShipmentTrackingPage.verifyShipmentDetails(shipmentId);
});

Given("shipment details for ID {string} are displayed", (shipmentId) => {
 ShipmentTrackingPage.verifyShipmentDetails(shipmentId);
});

When("the user checks the current status of the shipment", () => {
 ShipmentTrackingPage.checkCurrentStatus();
});

Then("the current status is displayed as {string}", (status) => {
 ShipmentTrackingPage.verifyCurrentStatus(status);
});

Given("the current status is {string}", (status) => {
 ShipmentTrackingPage.verifyCurrentStatus(status);
});

When("a status update to {string} is simulated", (newStatus) => {
 ShipmentTrackingPage.simulateStatusUpdate(newStatus);
});

Then("the status updates to {string} in real-time", (newStatus) => {
 ShipmentTrackingPage.verifyStatusUpdate(newStatus);
});

Given("the status is {string}", (status) => {
 ShipmentTrackingPage.verifyCurrentStatus(status);
});

When("the user verifies the timestamp of the latest status update", () => {
 ShipmentTrackingPage.verifyTimestamp();
});

Then("the timestamp matches the current time", () => {
 ShipmentTrackingPage.verifyTimestampMatchesCurrentTime();
});

Given("the status is {string} with the correct timestamp", (status) => {
 ShipmentTrackingPage.verifyCurrentStatus(status);
 ShipmentTrackingPage.verifyTimestampMatchesCurrentTime();
});

When("the user refreshes the page", () => {
 cy.reload();
});

Then("the status remains {string} with the same timestamp", (status) => {
 ShipmentTrackingPage.verifyCurrentStatus(status);
 ShipmentTrackingPage.verifyTimestampMatchesCurrentTime();
});

Given("the status is {string}", (status) => {
 ShipmentTrackingPage.verifyCurrentStatus(status);
});

When("the user logs out and logs back in", () => {
 cy.logout();
 cy.login();
});

Then("the shipment status remains {string}", (status) => {
 ShipmentTrackingPage.verifyCurrentStatus(status);
});

Given("the user is logged in", () => {
 cy.login();
});

When("the user checks the notification settings for shipment updates", () => {
 ShipmentTrackingPage.checkNotificationSettings();
});

Then("notifications are enabled for real-time updates", () => {
 ShipmentTrackingPage.verifyNotificationsEnabled();
});

Given("the user attempts to update status during a network issue", () => {
 ShipmentTrackingPage.simulateNetworkIssue();
});

When("the system handles the issue gracefully", () => {
 ShipmentTrackingPage.verifyNetworkIssueHandled();
});

Then("the system retries the update", () => {
 ShipmentTrackingPage.verifyRetryUpdate();
});

Given("the status updates have been made", () => {
 ShipmentTrackingPage.verifyStatusUpdatesMade();
});

When("the user verifies the shipment history log", () => {
 ShipmentTrackingPage.verifyHistoryLog();
});

Then("the log contains all status updates with timestamps", () => {
 ShipmentTrackingPage.verifyHistoryLogContents();
});

Given("the user updates the shipment status", () => {
 ShipmentTrackingPage.updateShipmentStatus();
});

When("the user checks for error messages", () => {
 ShipmentTrackingPage.checkForErrorMessages();
});

Then("no error messages are displayed", () => {
 ShipmentTrackingPage.verifyNoErrorMessages();
});

Given("the user has access to multiple devices", () => {
 ShipmentTrackingPage.verifyMultipleDeviceAccess();
});

When("the user attempts to update status from a different device", () => {
 ShipmentTrackingPage.updateStatusFromDifferentDevice();
});

Then("the status updates are synchronized across devices", () => {
 ShipmentTrackingPage.verifyStatusSynchronization();
});

Given("the user checks the shipment status on a mobile device", () => {
 ShipmentTrackingPage.checkStatusOnMobileDevice();
});

When("the status is displayed", () => {
 ShipmentTrackingPage.verifyStatusDisplayed();
});

Then("the status is consistent with the desktop view", () => {
 ShipmentTrackingPage.verifyStatusConsistency();
});

Given("the system has been rebooted", () => {
 ShipmentTrackingPage.rebootSystem();
});

When("the user checks the shipment status", () => {
 ShipmentTrackingPage.checkShipmentStatus();
});

Then("the status remains {string}", (status) => {
 ShipmentTrackingPage.verifyCurrentStatus(status);
});